pkg/api/fnresult/types.go

```diff
@@ -0,0 +1,47 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package fnresult
+
+import (
+	"sigs.k8s.io/kustomize/kyaml/fn/framework"
+	"sigs.k8s.io/kustomize/kyaml/yaml"
+)
+
+// FunctionResult contains the structured result from an individual function
+type FunctionResult struct {
+	// Image is the full name of the image that generates this result
+	// Image and Exec are mutually exclusive
+	Image string `yaml:"image,omitempty"`
+	// Exec is the the absolute path to the executable file
+	Exec string `yaml:"exec,omitempty"`
+	// ExitCode is the exit code from running the function
+	ExitCode int `yaml:"exitCode,omitempty"`
+	// Results is the list of results for the function
+	Results []framework.Item `yaml:"results,omitempty"`
+}
+
+const (
+	FunctionResultListKind       = "FunctionResultList"
+	FunctionResultListGroup      = "config.kubernetes.io"
+	FunctionResultListVersion    = "v1beta1"
```

Comment:
why not v1alpha2?

Also follow the directory convention e.g.

kpt/pkg/api/kptfile/v1alpha2/types.go

Reply:
> why not v1alpha2?

I cannot find a doc about what apiVersion should be used for List kinds so I copied the apiVersion from the KRM function spec. Does it mean the apiVersion for List kinds is not defined and we can choose what we want?

> Also follow the directory convention e.g.
> kpt/pkg/api/kptfile/v1alpha2/types.go

I didn't follow the pattern since I was not sure should we bind the function result list version with Kptfile version. Should we do this?

Reply:
You may be conflating two different things (I commented about this before):

1. The wire format as described in the Function Spec Standard.
2. The KRM resource outputted by the kpt CLI using --results flag. This is kpt-specific.

I assumed you're defining 2 here, not 1. These two do not necessarily need to be coupled to each other. `Results` field in `FunctonResult` is currently defiend as `framework.Item` for convinience, but they don't
have to be the same.

Reply:
Update the api version to v1alpha1

Reply:
Why?

As mentioned previously, this is a kpt-specific kind, so should use the same group and version as Kptfile: https://github.com/GoogleContainerTools/kpt/blob/next/pkg/api/kptfile/v1alpha2/types.go#L26

Reply:
Use the same group and version with Kptfile now.